fix(devtools): DOM traversal bug

AleksanderBodurri opened the pull request:
Previously, Angular devtools would mistakenly traverse the same DOM elements multiple times while doing traversal for the component tree explorer. This error case would occur when more than 1 Angular application root component was present on the same page and in distinct DOM branches.

An example of where it would enter the irregular behaviour (previous state)

```html
<app-root>
...
</app-root>
<other-app>
...
</other-app>
```
<img width="909" height="83" alt="Screenshot 2025-07-20 at 2 03 12 AM" src="https://github.com/user-attachments/assets/3ee17d9d-d393-4333-97b4-ab05b57412c0" />


Now, we properly ignore duplicate DOM paths when looking for application and non-application root component to begin the Angular DevTools component discovery logic.

<img width="904" height="44" alt="Screenshot 2025-07-20 at 2 04 45 AM" src="https://github.com/user-attachments/assets/d0508eda-28b1-4bdf-bd2b-5523e4dc5ab9" />

AleksanderBodurri commented on devtools/projects/ng-devtools-backend/src/lib/directive-forest/render-tree.spec.ts:
```diff
@@ -92,30 +92,6 @@ describe('render tree extraction', () => {
     expect(rtree[0].children[0].children.length).toBe(0);
   });
 
-  it('should go all the way to the root element to look up for nodes', () => {
```
The logic to traverse up to the DOM is no longer the responsibility of RTree strategy so this test is no longer necessary. 

The new behaviour is now tested for in `devtools/cypress/integration/node-search.e2e.js`.

JeanMeche commented on devtools/projects/ng-devtools-backend/src/lib/component-tree/component-tree.ts:
```diff
@@ -548,9 +548,102 @@ const getRootLViewsHelper = (element: Element, rootLViews = new Set<any>()): Set
+  }
+
+  const roots = getAppRoots();
+  const rootSet = new Set(roots);
+
+  // Traverse the DOM tree for other non application root Angular components.
+  // Pass in the existing roots to inform the traversal that we can skip those paths.
+  discoverNonApplicationRootComponents(document.body, rootSet);
+
+  return [...rootSet];
```
Nit: can we maybe move the whole `Set` part into the `discoverNonApplicationRootComponents` function. It has not real meanning here.

AleksanderBodurri replied:
discoverNonApplicationRootComponents is recursive with cases being determined by a `Set`, no easy way to do this without creating another helper which I don't think is necessarily better

JeanMeche commented on devtools/projects/ng-devtools-backend/src/lib/directive-forest/index.ts:
```diff
@@ -29,13 +30,33 @@ const selectStrategy = (element: Element) => {
+    strategy: RTreeStrategy | LTreeStrategy;
+    element: Element;
+  }[] = elements
+    .map((element) => {
```
Is this to support heterogenous versions on Angular ? If yes can we make this obvious via a comment.

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/directive-forest.component.ts:
```diff
@@ -165,7 +165,6 @@ export class DirectiveForestComponent {
     this.populateParents(node.position);
     this.selectNode.emit(node.original);
     this.selectedNode.set(node);
-    this.currentlyMatchedIndex.set(-1);
```
I think this was causing some unintended behaviour when paired with our new zoneless configuration. I'm not sure why currentlyMatchedIndex was getting set to -1 here at all. With this removed it seems like things are working as expected. Previously a select call would reset this index which would cause weird behaviour with the "next" and "prev" buttons

AleksanderBodurri committed:
fix(devtools): DOM traversal bug

Previously, Angular devtools would mistakenly traverse the same DOM elements multiple times while doing traversal for the component tree explorer. This error case would occur when more than 1 Angular application root component was present on the same page and in distinct DOM branches.

Some example cases that did work previously:

```html
<app-root>
...
</app-root>
```

```html
<app-root>
...
<app-root-2></app-root-2>
...
</app-root>
```

An example of where it would enter the irregular behaviour

```html
<app-root>
...
</app-root>
<app-root-2>
...
</app-root-2>
```

Now, we properly ignore duplicate DOM paths when looking for application and non-application root component to begin the Angular DevTools component discovery logic.

AleksanderBodurri commented on devtools/cypress/cypress.config.js:
```diff
@@ -7,6 +7,7 @@
  */
 
 module.exports = {
+  defaultCommandTimeout: 10000, // Increase the default command timeout to 10 seconds
```
This is a proactive bump because a slow enough CPU can cause test failures even if application behaviour is correct. This is not necessarily solving a particular test flake

The default is 4000ms, https://github.com/angular/angular/pull/62719/commits/834ac6b17433e3779ed0e6fe3a2e5ea7a454dd98

AleksanderBodurri replied:
I can split apart the last two commits into separate PRs if needed

AleksanderBodurri replied:
Moved into new PR

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/directive-forest/index.ts:
```diff
@@ -29,13 +30,35 @@ const selectStrategy = (element: Element) => {
+    if (strategy === rTreeStrategy) {
+      result.push(rTreeStrategy.build(element, index));
+    } else if (strategy === lTreeStrategy) {
+      result.push(lTreeStrategy.build(element));
```
Consider: Can we update `LTreeStrategy` and `RTreeStrategy` to use a shared interface (with `supports` and `build` accepting an index). Then `LTreeStrategy` can just ignore the index, but the caller can just always provide it. That way we don't need to disambiguate between the two here.

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/directive-forest/index.ts:
```diff
@@ -29,13 +30,35 @@ const selectStrategy = (element: Element) => {
+    }
+  }
+
+  return result.flat();
```
Consider: I think this `for` loop above could just be a `strategies.entries().map(() => ...)`?

dgp1130 reviewed (approved):
This seems like a good step in the right direction, so I think it's fine to move forward with this PR. I suspect there might be a future change we should consider in the framework so `ApplicationRef` actually has correct knowledge of all the root views in an application so we don't need to do this work in DevTools at all.

AleksanderBodurri commented on devtools/projects/ng-devtools-backend/src/lib/component-tree/component-tree.ts:
```diff
@@ -548,9 +548,102 @@ const getRootLViewsHelper = (element: Element, rootLViews = new Set<any>()): Set
+
+  // Traverse the DOM tree for other non application root Angular components.
+  // Pass in the existing roots to inform the traversal that we can skip those paths.
+  discoverNonApplicationRootComponents(document.body, rootSet);
```
Ahh looks like CdkOverlay and friends do collect in `ApplicationRef._views`. We don't use ApplicationRef to discover these roots though in Angular DevTools though. I think this warrants broader discussion around how to discover "Applications" through the ng global.

dgp1130 replied:
+1 to using this in a new `ng.getApps` / `ng.getRootNodes` implementation. Out of scope for this PR, but something we can consider in the future.

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/component-tree/component-tree.ts:
````diff
@@ -548,9 +548,104 @@ const getRootLViewsHelper = (element: Element, rootLViews = new Set<any>()): Set
+ * ```ts
+ * const rootSet = new Set(getAppRoots());
+ * console.log(rootSet);
+ * // Set(<app-root-1>, <aspp-root-2>)
````
Typo:

```suggestion
 * // Set(<app-root-1>, <app-root-2>)
```

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/directive-forest/index.ts:
```diff
@@ -29,13 +30,20 @@ const selectStrategy = (element: Element) => {
+      return [];
+    }
+
+    return strategy.build(element, i++);
```
Consider: I think `flatMap` provides an index argument you can use instead of tracking `i++` directly.

You might want to flatMap `element` to `[element, strategy]` first to handle the `!strategy` case first:

```typescript
elements.flatMap((element) => {
  const strategy = selectStrategy(element);
  if (!strategy) return [];

  return [element, strategy];
}).map(([element, strategy], index) => strategy.build(element, index));
```

AleksanderBodurri replied:
Returning an array from the map causes weird things to occur with the typing of element and strategy. They both end up being treated as `RTreeStrategy | LTreeStrategy | Element`. I think the current approach is sufficient to avoid this

dgp1130 reviewed (approved):
Couple minor suggestions to think about, but nothing too significant. Thanks for looking into this!
